File: Semester1/Fundamentals of Programming/seminarii/seminar12/merge_sort.py
"""
    @author: Badita Marin-Georgian
    @email:  geo.badita@gmail.com
    @date:   12/18/2017 09:13
"""
def merge(left, right, list):
    i = 0
    j = 0
    k = 0

    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            list[k] = left[i]
            i += 1
        else:
            list[k] = right[j]
            j += 1
        k += 1

    while i < len(left):
        list[k] = left[i]
        i += 1
        k += 1

    while j < len(right):
        list[k] = right[j]
        j += 1
        k += 1

def merge_sort(list):
    if len(list) > 1:
        mid = len(list)//2
        left = list[:mid]
        right = list[mid:]

        merge_sort(left)
        merge_sort(right)

        merge(left, right, list)


l = [5, 3, 4, 2, 1]
merge_sort(l)
print(l)

File: BachelorExamPrep/OOP/Execices/mergesort.cpp
//
// Created by geo on 17.06.2020.
//

#include "mergesort.h"

//arr = [1, 2, 3, 4, 5, 6];
//left = 0, right = 5, mid = 2;
//leftSide = [0, 1, 2]
//rightSide = [3, 4, 5]
void merge(int left, int mid, int right, int *arr) {
    int *leftSide = new int[mid - left + 1];
    int *rightSide = new int[right - mid];
    int lCount = 0;
    int rCount = 0;
    for (int i = left; i <= mid; i++) {
        leftSide[lCount++] = arr[i];
    }

    for (int i = mid + 1; i <= right; i++) {
        rightSide[rCount++] = arr[i];
    }
    int startPos = left;
    int i = 0, j = 0;
    while (i < lCount && j < rCount) {
        if (leftSide[i] < rightSide[j]) {
            arr[startPos] = leftSide[i];
            i++;
        } else {
            arr[startPos] = rightSide[j];
            j++;
        }
        startPos++;
    }

    while (i < lCount) {
        arr[startPos++] = leftSide[i++];
    }

    while (j < rCount) {
        arr[startPos++] = rightSide[j++];
    }

    delete[] leftSide;
    delete[] rightSide;
}

void mergeSort(int left, int right, int *arr) {
    if (left < right) {
        int mid = left + (right - left) / 2;
        mergeSort(left, mid, arr);
        mergeSort(mid + 1, right, arr);
        merge(left, mid, right, arr);
    }
}